[docs] [YB Voyager] Docs for live migration (#19025)
yugabyte/yugabyte-db: docs/content/preview/yugabyte-voyager/migrate/live-fall-forward.md

---
title: Steps to perform live migration of your database using YugabyteDB Voyager
headerTitle: Live migration with fall-forward
linkTitle: Live migration with fall-forward
headcontent: Steps for a live migration with fall-forward using YugabyteDB Voyager
description: Steps to ensure a successful live migration with fall-forward using YugabyteDB Voyager.
menu:
  preview_yugabyte-voyager:
    identifier: live-fall-forward
    parent: migration-types
    weight: 103
techPreview: /preview/faq/general/#what-is-the-definition-of-the-beta-feature-tag
rightNav:
  hideH4: true
type: docs
---

When you migrate a database using YugabyteDB Voyager, it is prudent to have a backup strategy if the new database doesn't work as expected. A fall-forward approach consists of creating a third database (Fall-forward database) that is a replica of your original source database.

A fall forward approach allows you to test the system end-to-end. This workflow is especially important in heterogeneous migration scenarios, in which source and target databases are using different engines.

## Fall-forward workflow

![fall-forward short](/images/migrate/live-fall-forward-short.png)

> After starting [live migration](../live-migrate/#live-migration-workflow) of data, an additional step of setting up the fall-forward database (via [fall-forward setup](#fall-forward-setup)) needs to be performed. The setup starts replicating the snapshot data along with new changes exported from the source database to the fall-forward DB, as shown in the following illustration:

Reviewer: ```suggestion
Before starting a [live migration](../live-migrate/#live-migration-workflow), you set up the fall-forward database (via [fall-forward setup](#fall-forward-setup)). During migration, yb-voyager replicates the snapshot data along with new changes exported from the source database to the target and fall-forward databases, as shown in the following illustration:
```

![After fall-forward setup](/images/migrate/after-fall-forward-setup.png)

At [cutover](#cutover), applications stop writing to the source database and start writing to the target YugabyteDB. After the cutover process is complete, YB Voyager keeps the fall-forward database synchronized with changes from the target Yugabyte DB as shown in the following illustration:

![After cutover](/images/migrate/after-cutover.png)

> Finally, if you must switch to the fall-forward database in case the current YugabyteDB system is not working as expected, you can use [fall-forward switchover](#fall-forward-switchover-optional) option to achieve the same.

Reviewer: ```suggestion
Finally, if you need to switch to the fall-forward database (because the current YugabyteDB system is not working as expected), you can use [fall-forward switchover](#fall-forward-switchover-optional) option.
```

![After fall-forward switchover](/images/migrate/after-fall-fwd-switchover.png)

> The following illustration describes the overview of live migration using YB Voyager with fall-forward option.

Reviewer: ```suggestion
The following illustration describes the workflow for live migration using YB Voyager with the fall-forward option.
```

![Live migration with fall-forward workflow](/images/migrate/live-fall-forward.png)

Before proceeding with migration, ensure that you have completed the following steps:

- [Install yb-voyager](../../install-yb-voyager/#install-yb-voyager).
- Check the [unsupported features](../../known-issues/#unsupported-features) and [known issues](../../known-issues/#known-issues).
- Review [data modeling](../../reference/data-modeling/) strategies.
- [Prepare the source database](#prepare-the-source-database).
- [Prepare the target database](#prepare-the-target-database).

## Prepare the source database

Prepare your source database by creating a new database user, and provide it with READ access to all the resources which need to be migrated.

<ul class="nav nav-tabs nav-tabs-yb">
  <li>

>     <a href="#standalone-oracle" class="nav-link" id="standalone-oracle-tab" data-toggle="tab" role="tab" aria-controls="oracle" aria-selected="true">

Reviewer: ```suggestion
    <a href="#standalone-oracle" class="nav-link active" id="standalone-oracle-tab" data-toggle="tab" role="tab" aria-controls="oracle" aria-selected="true">
```

yb-voyager import schema --export-dir <EXPORT_DIR> \
        --target-db-host <TARGET_DB_HOST> \
        --target-db-user <TARGET_DB_USER> \
        --target-db-password <TARGET_DB_PASSWORD> \ # Enclose the password in single quotes if it contains special characters..
        --target-db-name <TARGET_DB_NAME> \
        --target-db-schema <TARGET_DB_SCHEMA>
```

Refer to [import schema](../../reference/yb-voyager-cli/#import-schema) for details about the arguments.

yb-voyager applies the DDL SQL files located in the `$EXPORT_DIR/schema` directory to the target database. If yb-voyager terminates before it imports the entire schema, you can rerun it by adding the `--ignore-exist` option.

{{< note title="Importing indexes and triggers" >}}

Because the presence of indexes and triggers can slow down the rate at which data is imported, by default `import schema` does not import indexes and triggers. You should complete the data import without creating indexes and triggers. Only after data import is complete, create indexes and triggers using the `import schema` command with an additional `--post-import-data` flag.

{{< /note >}}

### Export and import schema to fall-forward database

> Manually, setup the fall-forward database with the same schema as that of the source database with the following considerations:

Reviewer: ```suggestion
Manually, set up the fall-forward database with the same schema as that of the source database with the following considerations:
```

> - The table names on the fall-forward database need to be case insensitive ( as YB Voyager currently does not support case-sensitivity.)

Reviewer: ```suggestion
- The table names on the fall-forward database need to be case insensitive (YB Voyager currently does not support case-sensitivity).
```

> - Do not create indexes and triggers at the schema setup stage, as it will degrade performance of importing data into the fall-forward database. Create them later as mentioned in Step 4 of [Fall-forward switchover](#fall-forward-switchover-optional).

Reviewer: ```suggestion
- Do not create indexes and triggers at the schema setup stage, as it will degrade performance of importing data into the fall-forward database. Create them later as described in [Fall-forward switchover](#fall-forward-switchover-optional).
```

| ---------------------------------------  |  ----------------------------- |
```

Note that the CDC phase will start only after a snapshot of the entire interested table-set is completed.
Additionally, the CDC phase is restartable. So, if yb-voyager terminates when data export is in progress, it resumes from its current state after the CDC phase is restarted.

#### Caveats

- Some data types are unsupported. For a detailed list, refer to [datatype mappings](../../reference/datatype-mapping-oracle/).
- For Oracle where sequences are not attached to a column, resume value generation is unsupported.
- [--parallel-jobs](../../reference/yb-voyager-cli/#parallel-jobs) argument has no effect on live migration.

Refer to [export data](../../reference/yb-voyager-cli/#export-data) for details about the arguments, and [export data status](../../reference/yb-voyager-cli/#export-data-status) to track the status of an export operation.

The options passed to the command are similar to the [`yb-voyager export schema`](#export-schema) command. To export only a subset of the tables, pass a comma-separated list of table names in the `--table-list` argument.

### Import data

> After you have successfully imported the schema in the target database, and the CDC phase has started in export data (which can be checked by export data status command), you can start importing the data using the yb-voyager import data command as follows:

Reviewer: ```suggestion
After you have successfully imported the schema in the target database, and the CDC phase has started in export data (which you can monitor using the export data status command), you can start importing the data using the yb-voyager import data command as follows:
```

| -----------------------------  |  ----------------------------- |
```

The entire import process is designed to be _restartable_ if yb-voyager terminates while the data import is in progress. If restarted, the data import resumes from its current state.

{{< note title="Note">}}
[table-list](../../reference/yb-voyager-cli/#table-list) and [exclude-table-list](../../reference/yb-voyager-cli/#exclude-table-list) flags are not supported in live migration.
{{< /note >}}

{{< tip title="Importing large datasets" >}}

When importing a very large database, run the import data command in a `screen` session, so that the import is not terminated when the terminal session stops.

If the `yb-voyager import data` command terminates before completing the data ingestion, you can re-run it with the same arguments and the command will resume the data import operation.

{{< /tip >}}

#### Import data status

Run the `yb-voyager import data status --export-dir <EXPORT_DIR>` command to get an overall progress of the data import operation.

### Fall-forward setup

Note that the fall-forward setup is applicable for data migration only (schema migration needs to be done manually).

> The fall-forward setup refers to replicating the snapshot data along with the changes exported from the source database to the fall-forward database as shown in the "After Cutover" illustration . The command to start the setup as follows:

Reviewer: ```suggestion
The fall-forward setup refers to replicating the snapshot data along with the changes exported from the source database to the fall-forward database. The command to start the setup is as follows:
```

```sh
yb-voyager fall-forward setup --export-dir <EXPORT-DIR> \
--ff-db-host <HOST> \
--ff-db-user <USERNAME> \
--ff-db-password <PASSWORD> \ # Enclose the password in single quotes if it contains special characters.
--ff-db-name <DB-NAME> \
--ff-db-schema <SCHEMA-NAME> \
--parallel-jobs <COUNT>
```

Refer to [fall-forward setup](../../reference/yb-voyager-cli/#fall-forward-setup-tech-preview) for details about the arguments.

Similar to [import data](#import-data), during fall-forward:

- The snapshot is first imported, following which, the change events are imported to the fall-forward database.

> - Some important metrics such as the number of events, events rate, and so on is displayed.

Reviewer: ```suggestion
- Some important metrics such as the number of events, events rate, and so on, are displayed.
```

- The setup is restartable.

> Additionally, when you run the `fall-forward setup` command, the [import data status](#import-data-status) command also shows progress of importing all changes to the fall-forward database to get an overall progress of the data import operation and streaming changes to fall-forward DB with the following command:

Reviewer: ```suggestion
Additionally, when you run the `fall-forward setup` command, the [import data status](#import-data-status) command also shows progress of importing all changes to the fall-forward database. To view overall progress of the data import operation and streaming changes to the fall-forward database, use the following command:
```

1. Verify your migration. After the schema and data import is complete, the automated part of the database migration process is considered complete. You should manually run validation queries on both the source and target database to ensure that the data is correctly migrated. A sample query to validate the databases can include checking the row count of each table.

    {{< warning title = "Caveat associated with rows reported by import data status" >}}

Suppose you have a scenario where,

- [import data](#import-data) or [import data file](#import-data-file) command fails.
- To resolve this issue, you delete some of the rows from the split files.
- After retrying, the import data command completes successfully.

In this scenario, [import data status](#import-data-status) command reports incorrect imported row count; because it doesn't take into account the deleted rows.

For more details, refer to the GitHub issue [#360](https://github.com/yugabyte/yb-voyager/issues/360).

    {{< /warning >}}

1. Stop [archive changes](#archive-changes).

### Fall-forward switchover (Optional)

> During this phase, switch over your application from pointing it to the target YugabyteDB to pointing it to the fall-forward database. As this step is optional, perform it _only_ if the target YugabyteDB is not working as expected.

Reviewer: ```suggestion
During this phase, switch over your application from the target YugabyteDB database to the fall-forward database. As this step is optional, perform it _only_ if the target database is not working as expected.
```

Keep monitoring the metrics displayed for `fall-forward synchronize` and `fall-forwad setup` processes. After you notice that the import of events to the fall-forward database is catching up to the exported events from the target database, you are ready to cutover. You can use the "Remaining events" metric displayed in the import data process to help you determine the cutover.

Perform the following steps as part of the cutover process:

1. Quiesce your source database, that is stop application writes.

> 1. Initiate switchover as follows:

Reviewer: single step

>     1. After the exported events rate ("ingestion rate" in the metrics table) drops to 0, it is safe to switchover using the following command:

Reviewer: ```suggestion
    1. After the exported events rate ("ingestion rate" in the metrics table) drops to 0, it is safe to switch over using the following command:
```

> 

Reviewer: cut over to a database
switch over your database
perform a cutover
perform a switchover
also, why are we using both? can clean this up later though

Author: Switchover is needed only if the target yugabyteDB database is not working as expected. So its an option for such a case.

Reviewer: This was just a note on using the terms, that is, 'cut over' when a verb, and 'cutover' when a noun
So we are distinguishing between cutover from Oracle to YBDB, and switchover from YBDB to fall forward YBDB?

Author: Got it. Thanks!

        ```sh
        yb-voyager fall-forward switchover --export-dir <EXPORT_DIR>
        ```

        Refer to [fall-forward switchover](../../reference/yb-voyager-cli/#fall-forward-switchover-tech-preview) for details about the arguments.

        The `fall-forward switchover` command stops the `fall-forward synchronize` process, followed by the `fall-forward setup` process after it has imported all the events to the fall-forward database.

> 1. Wait for the switchover process to complete. The status of the switchover process can be monitored by the following command:

Reviewer: ```suggestion
1. Wait for the switchover process to complete. Monitor the status of the switchover process using the following command:
```
